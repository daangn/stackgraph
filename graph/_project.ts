import { Reducer, Stream } from "../deps/rimbu.ts"
import { Project, SourceFile } from "../deps/ts_morph.ts"
import { denoCompilerOptions } from "../utils/project.ts"

/** In-Memory ESM Project for quick testing */
export const inMemoryProject = () =>
	new Project({
		compilerOptions: denoCompilerOptions,
		useInMemoryFileSystem: true,
	})

/**
 * Add source files to in-memory project
 */
export const withSrc =
	(project: Project) =>
	<const T extends string>(src: Record<T, string>): Record<T, SourceFile> =>
		Stream.fromObject(src)
			.map(([path, text]) =>
				[path, project.createSourceFile(path, text)] as const
			)
			.reduce(Reducer.toJSObject())
