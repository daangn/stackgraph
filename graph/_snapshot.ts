const read = await Deno.permissions.query({ name: "read" })

const ignore = read.state !== "granted"

/**
 * same as `Deno.test` but skipped if `read` permission is denied
 */
export const snapshotTest = (
	name: Deno.TestDefinition["name"],
	fn: Deno.TestDefinition["fn"],
) => Deno.test({ name, ignore, fn })
