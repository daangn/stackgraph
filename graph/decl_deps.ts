import { Stream } from "../deps/rimbu.ts"
import {
	type ClassDeclaration,
	type FunctionDeclaration,
	type Node,
	ReferencedSymbol,
	ReferencedSymbolEntry,
	SyntaxKind,
	type VariableDeclaration,
} from "../deps/ts_morph.ts"
import { encodeVSCodeURI } from "./vscode_uri.ts"

export type Declaration =
	| VariableDeclaration
	| ClassDeclaration
	| FunctionDeclaration

export type DeclDeps = Map<Declaration, Declaration[]>

/**
 * Get the topmost declaration
 *
 * @privateRemarks
 *
 * FIXME(#1): handle destructuring and multiple variable declarations
 * RFC: only track top-level declarations or inside if statements?
 * FIXME: also track `function` statements, are there other types of declarations?
 */
export const getTopDecl = (
	ref: ReferencedSymbolEntry,
): Declaration | undefined => {
	const target = ref.getNode().getAncestors()

	// 	.at(-2) // HACK: last is SourceFile, second to last is the variable/class declaration

	// return (target?.isKind(SyntaxKind.VariableDeclaration) ||
	// 		target?.isKind(SyntaxKind.ClassDeclaration))
	// 	? target
	// 	: undefined

	return target.findLast((x): x is Declaration =>
		x.isKind(SyntaxKind.VariableDeclaration) ||
		x.isKind(SyntaxKind.ClassDeclaration) ||
		x.isKind(SyntaxKind.FunctionDeclaration)
	)
}

export const equals = (a: Node, b: Node) =>
	encodeVSCodeURI(a) === encodeVSCodeURI(b)

export const getActualReferences = (
	symbol: ReferencedSymbol,
): Stream<ReferencedSymbolEntry> =>
	Stream.from(symbol
		.getReferences())
		.filter((ref) => !equals(ref.getNode(), symbol.getDefinition().getNode()))
		.filter((ref) =>
			ref.getNode().getParent()?.getKindName() !== "ImportClause"
		)

export const getReferencedDecls = (
	node: Declaration,
): Stream<Declaration> =>
	Stream.from(node.findReferences())
		.flatMap(getActualReferences)
		.map((x) => getTopDecl(x))
		.filter((x): x is Declaration => x !== undefined)

export const getDeclDeps = (links: Declaration[]): DeclDeps => {
	const graph: DeclDeps = new Map()

	let current = links
	while (current.length > 0) {
		current = current.flatMap((node) => {
			const references = getReferencedDecls(node).toArray()
			graph.set(node, references)
			return references
		})
	}
	return graph
}
