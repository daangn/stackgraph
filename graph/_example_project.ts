import { Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { inMemoryProject } from "./_project.ts"

/**
 * https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements
 *
 * Example file of all top-level javascript declarations
 */
export const declExampleText = (h: string | number = "") => /*javascript*/ `
    // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements#declaring_variables
    export var varVar${h} = 1
    export var varVarIgnore${h} = 2

    export let varLet${h} = 1
    export let varLetIgnore${h} = 2

    export const varConst${h} = 1
    export const varConstIgnore${h} = 2

    export const varArrowFn${h} = () => {}
    export const varArrowFnIgnore${h} = () => {}

    // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements#functions_and_classes
    export function functionFn${h}() {}
    export function functionFnIgnore${h}() {}

    export function${h}* functionGen() {}
    export function${h}* functionGenIgnore() {}

    export async function${h} functionAsync() {}
    export async function${h} functionAsyncIgnore() {}

    export async function${h}* functionAsyncGen() {}
    export async function${h}* functionAsyncGenIgnore() {}

    export class Class${h} {}
    export class ClassIgnore${h} {}

    // others
    if (true) {
        const variableInIf${h} = 3
    }
`

/**
 * Arbitarily large file with many declarations for benchmarking
 */
export const getDeclExampleFile = (amount: number) => {
	const project = inMemoryProject()

	const text = Stream
		.range({ amount }).map(declExampleText).join({ sep: "\n\n" })

	return project.createSourceFile("a.ts", text)
}

export const exampleSrc = {
	"a.ts": /*javascript*/ `
        export const a = "a"

	    export const aaa = () => a
	`,
	"b.tsx": /*javascript*/ `
        import { a, aaa } from "./a.ts"
        export { a }

        export const Comp = () => <div>{a}</div>
        export const CompAAA = () => <div>{aaa()}</div>

        // unrelated to a and should be ignored
        export const Unrelated = () => <div>Unrelated</div>
    `,
	"c.tsx": /*javascript*/ `
        import { Comp } from "./b.tsx"

        export const Page = () => <div><Comp/></div>
    `,
	"d.tsx": /*javascript*/ `
        import { a } from "./b.tsx"
        import { Page } from "./c.tsx"

        export const InnerImport = () => <Page/>
        export const AliasedImport = () => <div>{a}</div>
    `,
}
