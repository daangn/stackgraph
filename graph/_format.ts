// 1.40.5: https://github.com/denoland/deno/issues/22496
import { stripAnsiCode } from "https://deno.land/std@0.216.0/fmt/colors.ts"
import { Reducer, Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { Declaration, DeclDeps } from "./decl_deps.ts"
import { encodeVSCodeURI, prettyPrintURI } from "./vscode_uri.ts"

export const serializeNoColor = (x: unknown) =>
	stripAnsiCode(
		Deno.inspect(x, {
			depth: Infinity,
			colors: false,
			sorted: true,
			trailingComma: true,
			compact: false,
			iterableLimit: Infinity,
			breakLength: Infinity,
			escapeSequences: false,
			strAbbreviateSize: Infinity,
		}),
	)

export const asRecord =
	<T extends string | number | symbol>(fn: (decl: Declaration) => T) =>
	(declDeps: DeclDeps): Record<T, Set<T>> =>
		Stream.from(declDeps.entries())
			.map(([decl, deps]) => [fn(decl), new Set(deps.map(fn))] as [T, Set<T>])
			.reduce(Reducer.toJSObject())

export const declDepsSerializer = (declDeps: DeclDeps) =>
	serializeNoColor(
		asRecord((x) => prettyPrintURI(encodeVSCodeURI(x)))(declDeps),
	)
