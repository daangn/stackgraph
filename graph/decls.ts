import {
	Stream,
	StreamSource,
} from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { SourceFile } from "../deps/ts_morph.ts"
import { Declaration } from "./decl_deps.ts"

/**
 * Find initial declarations to begin analyze with.
 */
export const getDecls =
	(filter: (node: Declaration) => boolean) =>
	(files: StreamSource<SourceFile>): Stream<Declaration> =>
		Stream.from(files).flatMap((file) =>
			Stream.from<Declaration>(
				file.getVariableDeclarations(),
				file.getFunctions(),
				file.getClasses(),
			).filter(filter)
		)

export const asMetadataRecord = <A, B>(links: LinkedDecl<A, B>[]) =>
	Object.fromEntries(
		links.map(({ node, ...metadata }) => [node.getName()!, metadata]),
	)
