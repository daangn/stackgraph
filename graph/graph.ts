import type { GraphElement } from "https://deno.land/x/rimbu@1.2.0/graph/custom/common/link.ts"
import { ArrowGraphHashed } from "https://deno.land/x/rimbu@1.2.0/graph/mod.ts"
import { Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { DeclDeps } from "./decl_deps.ts"
import { encodeVSCodeURI, type VSCodeURI } from "./vscode_uri.ts"

/** directed graph of references */
export type Graph = ArrowGraphHashed<VSCodeURI>

export const declDepsToGraph = (declDeps: DeclDeps): Graph =>
	Stream.from(declDeps).flatMap(([source, targets]) =>
		Stream.from(targets).map((to): GraphElement<VSCodeURI> => [
			encodeVSCodeURI(source),
			encodeVSCodeURI(to),
		])
	)
		.reduce(ArrowGraphHashed.reducer())
