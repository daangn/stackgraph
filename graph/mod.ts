import { distinct } from "https://deno.land/std@0.215.0/collections/distinct.ts"
import {
	SourceFile,
	VariableDeclaration,
} from "https://deno.land/x/ts_morph@21.0.1/mod.ts"
import { getGraph } from "./graph.ts"
import type { Graph } from "./graph.ts"

export type Relations<A, B> = Record<string, Omit<LinkedVarDecl<A, B>, "decl">>
export type LinkedVarDecl<A, B> = {
	decl: VariableDeclaration
	links: A
	metas: B
}

export type Metadata = {
	logShowEvents: string[]
	logClickEvents: string[]
}

export type Dependencies = Record<string, string[]>

export type Rels = Record<string, Rel>
export type Rel = {
	links: Record<"push" | "replace", string[]>
	meta: Metadata
}

/**
 * creates a reference graph
 */
export const serialize = (graph: Graph) =>
	Array.from(graph.entries()).map(([decl, refs]) =>
		[decl.getName(), distinct(refs.map((x) => x.getName()))] as const
	)

type Option<A, B> = {
	/**
	 * Filter and map links, metadatas, and variable declaration
	 */
	getLink: (node: VariableDeclaration) => LinkedVarDecl<A, B> | undefined
}

export const generateGraph = <A, B>({ getLink }: Option<A, B>) =>
(
	files: SourceFile[],
): { dependencies: Dependencies; relations: Relations<A, B> } => {
	const links = files.flatMap((file) =>
		file.getVariableDeclarations().flatMap((decl) => getLink(decl) ?? [])
	)

	const relations = Object.fromEntries(
		links.map(({ decl, ...relations }) => [decl.getName(), relations]),
	)
	const dependencies = Object.fromEntries(
		links.flatMap(({ decl }) => serialize(getGraph(decl))),
	)

	return { dependencies, relations }
}
