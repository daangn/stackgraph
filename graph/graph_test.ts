import { Reducer, Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { exampleSrc } from "./_example_project.ts"
import { inMemoryProject, withSrc } from "./_project.ts"
import { declDepsToGraph, Graph } from "./graph.ts"
import { getAllDecls } from "./decls.ts"
import { getDeclDeps } from "./decl_deps.ts"
import { assertSnapshot } from "../test_deps.ts"
import { parseVSCodeURI, prettyPrintURI } from "./vscode_uri.ts"
import { serializeNoColor } from "./_format.ts"
import { snapshotTest } from "./_snapshot.ts"

const project = inMemoryProject()
const files = withSrc(project)(exampleSrc)
const decls = Stream.fromObjectValues(files).flatMap(getAllDecls).toArray()
const declDeps = getDeclDeps(decls)
const graph = declDepsToGraph(declDeps)

/**
 * 1. groups by source file
 * 2. pretty-print connections
 */
const prettyPrintGraph = (graph: Graph) => {
	const groups = Object.groupBy(
		graph.streamNodes(),
		(node) => parseVSCodeURI(node).path,
	)

	return Stream.fromObject(groups)
		.map(([path, nodes]) => {
			const values = Stream.from(nodes).map((node) => {
				const connections = graph
					.getConnectionsFrom(node)
					.stream().map(prettyPrintURI).toArray()

				return [prettyPrintURI(node), connections] as const
			})
				.reduce(Reducer.toJSObject())
			return [path, values] as const
		})
		.reduce(Reducer.toJSObject())
}

export const graphSerializer = (graph: Graph) =>
	serializeNoColor(prettyPrintGraph(graph))

snapshotTest(
	"declDepsToGraph() converts declDeps into valid graph",
	async (t) => {
		await assertSnapshot(t, graph, { serializer: graphSerializer })
	},
)
