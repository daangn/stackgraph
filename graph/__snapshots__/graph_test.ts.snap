export const snapshot = {};

snapshot[`declDepsToGraph() converts declDeps into valid graph 1`] = `
{
  "/a.ts": {
    "a (/a.ts:2:22)": [
      "AliasedImport (/d.tsx:6:22)",
      "Comp (/b.tsx:5:22)",
      "aaa (/a.ts:4:19)",
    ],
    "aaa (/a.ts:4:19)": [
      "CompAAA (/b.tsx:6:22)",
    ],
  },
  "/b.tsx": {
    "Comp (/b.tsx:5:22)": [
      "Page (/c.tsx:8:22)",
    ],
    "CompAAA (/b.tsx:6:22)": [],
  },
  "/c.tsx": {
    "Page (/c.tsx:8:22)": [
      "InnerImport (/d.tsx:5:22)",
    ],
  },
  "/d.tsx": {
    "AliasedImport (/d.tsx:6:22)": [],
    "InnerImport (/d.tsx:5:22)": [],
  },
}
`;
